op-node/rollup/derive/plasma_data_source_test.go

```diff
@@ -121,4 +178,281 @@ func TestPlasmaDataSource(t *testing.T) {
+		}
+
+	}
+	finalitySignal.AssertExpectations(t)
+	l1F.AssertExpectations(t)
+}
+
+// This tests makes sure the pipeline returns a temporary error if data is not found.
+func TestPlasmaDataSourceStall(t *testing.T) {
+	logger := testlog.Logger(t, log.LevelDebug)
+	ctx := context.Background()
+
+	rng := rand.New(rand.NewSource(1234))
+
+	l1F := &testutils.MockL1Source{}
+
+	storage := plasma.NewMockDAClient(logger)
+
+	pcfg := plasma.Config{
+		ChallengeWindow: 90, ResolveWindow: 90,
+	}
+
+	da := plasma.NewPlasmaDAWithStorage(logger, pcfg, storage, l1F, &plasma.NoopMetrics{})
+
+	finalitySignal := &MockFinalitySignal{}
+	da.OnFinalizedHeadSignal(finalitySignal.OnNewL1Finalized)
+
+	// Create rollup genesis and config
+	l1Time := uint64(2)
+	refA := testutils.RandomBlockRef(rng)
+	refA.Number = 1
+	l1Refs := []eth.L1BlockRef{refA}
+	refA0 := eth.L2BlockRef{
+		Hash:           testutils.RandomHash(rng),
+		Number:         0,
+		ParentHash:     common.Hash{},
+		Time:           refA.Time,
+		L1Origin:       refA.ID(),
+		SequenceNumber: 0,
+	}
+	batcherPriv := testutils.RandomKey()
+	batcherAddr := crypto.PubkeyToAddress(batcherPriv.PublicKey)
+	batcherInbox := common.Address{42}
+	cfg := &rollup.Config{
+		Genesis: rollup.Genesis{
+			L1:     refA.ID(),
+			L2:     refA0.ID(),
+			L2Time: refA0.Time,
+		},
+		BlockTime:         1,
+		SeqWindowSize:     20,
+		BatchInboxAddress: batcherInbox,
+		UsePlasma:         true,
+	}
+
+	signer := cfg.L1Signer()
+
+	factory := NewDataSourceFactory(logger, cfg, l1F, nil, da)
+
+	parent := l1Refs[0]
+	// create a new mock l1 ref
+	ref := eth.L1BlockRef{
+		Hash:       testutils.RandomHash(rng),
+		Number:     parent.Number + 1,
+		ParentHash: parent.Hash,
+		Time:       parent.Time + l1Time,
+	}
+	l1F.ExpectFetchReceipts(ref.Hash, nil, types.Receipts{}, nil)
+	// mock input commitments in l1 transactions
+	input := testutils.RandomData(rng, 2000)
+	comm, _ := storage.SetInput(ctx, input)
+
+	tx, err := types.SignNewTx(batcherPriv, signer, &types.DynamicFeeTx{
+		ChainID:   signer.ChainID(),
+		Nonce:     0,
+		GasTipCap: big.NewInt(2 * params.GWei),
+		GasFeeCap: big.NewInt(30 * params.GWei),
+		Gas:       100_000,
+		To:        &batcherInbox,
+		Value:     big.NewInt(int64(0)),
+		Data:      comm,
+	})
+	require.NoError(t, err)
+
+	txs := []*types.Transaction{tx}
+
+	l1F.ExpectInfoAndTxsByHash(ref.Hash, testutils.RandomBlockInfo(rng), txs, nil)
+
+	// delete the input from the DA provider so it returns not found
+	require.NoError(t, storage.DeleteData(comm))
+
+	// next block is fetched to look ahead challenges but is not yet available
+	l1F.ExpectL1BlockRefByNumber(ref.Number+1, eth.L1BlockRef{}, ethereum.NotFound)
+
+	src, err := factory.OpenData(ctx, ref, batcherAddr)
+	require.NoError(t, err)
+
+	// data is not found so we return a temporary error
+	_, err = src.Next(ctx)
+	require.ErrorIs(t, err, ErrTemporary)
+
+	// now challenge is resolved
+	da.State().SetResolvedChallenge(comm, input, ref.Number+2)
```

Comment:
Do we have a test which has the `damgr` load in the resolved challenge from the transactions rather than manually setting the challenge to be resolved?